### PR TITLE
Euler incident updates

### PR DESCRIPTION
**Pull request by bugduino**
This PR won't be closed as is. This is used only to upgrade IdleCDOs affected by Euler Hack.

IdleCDO.sol was restored from commit `1957074706ce0ec2bf8ea7b92fb93c3e101f55b3` (ie remove the automatic loss mgmt as it was not audited yet).
A `setRedemptionRates` was added to distribute losses among tranche holders.
To run test comment the following files:
`test/foundry/IdleCDOLossMgmt.t.sol`
`test/foundry/IdleLeveragedEulerStrategy.t.sol`

### CODE REVIEW

File: contracts/IdleCDO.sol [hunk 1]
```diff
@@ -941,6 +941,37 @@ contract IdleCDO is PausableUpgradeable, GuardedLaunchUpgradable, IdleCDOStorage
...
+    if (redeemPrice > _lastAAPrice) {
+      // if yes set it to the last saved price
+      redeemPrice = _lastAAPrice;
+      // calculate funds for AA and BB
+      uint256 fundsAA = _AASupply * redeemPrice / ONE_TRANCHE_TOKEN;
+      uint256 fundsBB = _underlyingBalance - fundsAA;
+      // set final redeem prices
+      priceBB = fundsBB * ONE_TRANCHE_TOKEN / IERC20(BBTranche).totalSupply();
+    } else {
+      priceBB = 0;
```
Comment (massun-onibakuchi) [inline contracts/IdleCDO.sol:970]: memo: in this scenario, BB holder won't receive any funds.

File: contracts/IdleCDO.sol [hunk 2]
```diff
@@ -941,6 +941,37 @@ contract IdleCDO is PausableUpgradeable, GuardedLaunchUpgradable, IdleCDOStorage
...
+
+    address _AATranche = AATranche;
+    uint256 _AASupply = IERC20(_AATranche).totalSupply();
+    uint256 _lastAAPrice = _tranchePrice(_AATranche);
+    // Get all underlyings in the contract
+    uint256 _underlyingBalance = IERC20(token).balanceOf(address(this));
+    // Set redeem price as if everything is going to AA 
+    uint256 redeemPrice = _underlyingBalance * ONE_TRANCHE_TOKEN / _AASupply;
+    // Check if rate is too high compared to the last saved price
+    if (redeemPrice > _lastAAPrice) {
```
Comment (massun-onibakuchi) [inline contracts/IdleCDO.sol:961]: memo: in this scenario, AA holder would receive underlying proportional to lastAAPrice.

**Review by massun-onibakuchi — COMMENTED** [review]
looks good

**Comment by bugduino** [comment]
Things to note:
- These updates won't be merged on master, those will be dropped once redeems have been performed
- `IdleCDO`, `IdleEulerStrategy` and `IdleEulerStakingStrategy` have been updated to remove calls to eToken contracts (IdleCDO will be updated only for PYTs that do not use euler staking)
- `IdleCDO` as noted in the PR comment was restored to a version which has not automatic loss mgmt (because it was not yet audited) and the only change from there is the update of `getContractValue` to remove eTokens calls
- On strategies I commented out a lot of code so to have strategy price and apr == 0 and commented also redeem methods as those won't be used (we already have funds to send to IdleCDO)
- 2 tests `test/foundry/IdleCDOLossMgmt.t.sol` and `test/foundry/IdleLeveragedEulerStrategy.t.sol` have been commented out as those needs the master version of IdleCDO
- main test is `test/foundry/TestIdleCDODefaulMgmt.sol` where I tested the redeems for all affected PYTs and BYs

**Pull request closed without merging**